h2/src/main/org/h2/store/fs/FileBaseDefault.java

```diff
@@ -0,0 +1,58 @@
+/*
+ * Copyright 2004-2019 H2 Group. Multiple-Licensed under the MPL 2.0,
+ * and the EPL 1.0 (https://h2database.com/html/license.html).
+ * Initial Developer: H2 Group
+ */
+package org.h2.store.fs;
+
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.nio.channels.FileChannel;
+
+/**
+ * Default implementation of the slow operations that need synchronisation because they
+ * involve the file position.
+ */
+public abstract class FileBaseDefault extends FileBase {
+
+    private long position = 0;
+    
+    @Override
+    public synchronized final long position() throws IOException {
+        return 0;
+    }
+    
+    @Override
+    public final synchronized FileChannel position(long newPosition) throws IOException {
+        if (newPosition < 0) {
+            throw new IllegalArgumentException();
+        }
+        position = newPosition;
+        return this;
+    }
+    
+    @Override
+    public final synchronized int read(ByteBuffer dst) throws IOException {
+        int read = read(dst, position);
```

Comment:
Why that read() has to be under lock? We only need to protect usage of the "position" field (or even make it atomic instead) and let actual read() / write() concurrency to be taken care of by underlying FileChannel.

Reply:
(1) this read/write is only rarely used, so is not performance critical
(2) depending on the specific subclass, may not have something underneath it that is synchronised.
(3) we don't know exactly how much will actually be read/written, since the subclass may read/write less than requested, so we need to lock over the whole operation.

Reply:
Sorry, I missed that item (3). Thanks!